prototype/galore/galore_fused/README.md

```diff
@@ -0,0 +1,211 @@
+```
+
+#### Discussion
+
+##### Down Projection GEMM Shape
+
+The motivation for the `hybrid` approach is the unconventional matrix shapes of the down projection (Step 1):
+
+- The projection is always done such that the larger dimension of the `grad` matrix is maintained while other is projected to low rank per the `GaLore` algorithm
+  - E.g., if `M >= N`, the GEMM is of shape (`M x N`) x (`N x rank`) = (`M x rank`), (`rank x M`) x (`M x N`) = (`rank x N`) otherwise
+- Since `{M, N} >> rank` by definition, this results in a large reduction dimension relative to one of the output dimensions (output matrix is either fat or skinny)
+- This does not fit cleanly into the `split-k / parallel reduction` `GEMM` paradigm which is more tailored for shapes where both output dims are smaller than the reduction dimension.
+- Consequently, I had trouble finding an optimal kernel config using `triton` `autotuner` for the down projection step, despite tuning across many compute and io-bound configs (see `fused.triton_utils.kernels.matmul.py`).
+- Benchmarking `triton`-tuned `matmul` against default `torch.matmul` for these shapes showed worse performance, for `torch.float32`
+
+#### Effect of `TF32` tensor cores
+
+`allow_tf32`: this has significant impact on relative performance of `triton` vs `torch` matmuls:
+
+- Quick benchmarks of the downprojection `matmul` show that:
+  - with `allow_tf32=True` for both, triton exhibits `~1.30x` performance improvement over `torch`.
+  - with `allow_tf32=False`, performance of `triton` degrades significantly to `~.50x` of `torch`.
+
+See this [`torch note`](https://pytorch.org/docs/stable/notes/cuda.html#tf32-on-ampere) for more details on this feature.
+
+**Note**: This might be less of a concern given this incoming triton [PR](https://github.com/openai/triton/pull/3234), which implements a fast `TF32` trick that improves both performance and accuracy.
+
+#### Repro
+
+`tests/test_fused_kernels.py` is a `CLI` that has 2 modes, one for testing kernel accuracy, and the other for benchmarking across a number of configs.
+
+**Examples**
+
+_Accuracy_
+
+- Test accuracy of `torch` vs `hybrid` for `M=4096`, `N=4096`, `rank=128`, and `tf32` switched on:
+
+  ```python
+  python tests/test_fused_kernels.py --mode=test --kernel=hybrid --M=4096 --N=4096 --rank=128 --allow_tf32
+  ```
+
+_Benchmark_
+
+- Benchmark across all kernels without `tf32`:
+
+  ```python
+  python tests/test_fused_kernels.py --mode=benchmark
+  ```
+
+_Additional options_
+
+```python
+  python tests/test_fused_kernels.py --help
+```
+
+_Note:_ Passing in the additional flag `--verbose` will show `triton` autotuning logs -- I customized the `triton` autotuner spit out configs and other details.
+
+#### Test Env
+
+- GPU Device Props:
+  - Name: `NVIDIA RTX A6000`
+  - CC: `86`
+  - Total_memory: `48676MB`
+  - SM count: `84`
+- Torch: `2.2.2`
+- Triton: `2.2.0`
+
+#### Next Steps
+
+- [ ] Implement `FusedGaLoreOptimizer`
```

Comment:
Nit: Generally for next steps I'd rather they get mentioned in a github issue vs docs